datafusion/physical-plan/src/aggregates/group_values/single_group_by/primitive.rs

```diff
@@ -156,23 +183,33 @@ where
     }
 
     fn len(&self) -> usize {
-        self.values.len()
+        self.values.len().saturating_sub(1) * BLOCK_SIZE
+            + self.values.last().map_or(0, |v| v.len())
     }
 
     fn emit(&mut self, emit_to: EmitTo) -> Result<Vec<ArrayRef>> {
```

Comment:
Emit could be changed to support returning multiple arrays (`Result<Vec<Vec<ArrayRef>>>`)